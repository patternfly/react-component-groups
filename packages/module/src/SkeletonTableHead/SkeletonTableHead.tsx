import type { FunctionComponent, ReactElement } from 'react';
import { ReactNode, useMemo, isValidElement } from 'react';
import {
  Th,
  ThProps,
  Thead,
  Tr
} from '@patternfly/react-table';
import { Skeleton } from '@patternfly/react-core';

export const isThObject = (value: ReactNode | { cell: ReactNode; props?: ThProps }): value is { cell: ReactNode; props?: ThProps } => value != null && typeof value === 'object' && 'cell' in value;

export const isReactElement = (value: ReactNode): value is ReactElement => isValidElement(value);

export interface SkeletonTableHeadProps {
  /** Custom columns for the table */
  columns?: (ReactNode | { cell: ReactNode; props?: ThProps })[];
  /** Number of columns in the table */
  columnsCount?: number;
  /** Custom OUIA ID */
  ouiaId?: string | number;
  /** Flag indicating if the table is selectable */
  isSelectable?: boolean;
  /** Flag indicating if the table is expandable */
  isExpandable?: boolean;
  /** Flag indicating if the table is a tree table */
  isTreeTable?: boolean;
}

export const SkeletonTableHead: FunctionComponent<SkeletonTableHeadProps> = ({
  ouiaId = 'SkeletonTableHeader',
  isSelectable,
  isExpandable,
  columnsCount,
  columns,
  isTreeTable,
  ...rest
}: SkeletonTableHeadProps) => {
  const hasCustomColumns = Array.isArray(columns);
  const rowCellsCount = hasCustomColumns ? columns.length : columnsCount;

  const cells = useMemo(() => [
    ...(isExpandable ? [ <Th key="row-expand" screenReaderText='Data expansion table header cell' /> ] : []),
    ...(isSelectable && !isTreeTable ? [ <Th key="row-select" screenReaderText='Data selection table header cell' /> ] : []),
    ...(hasCustomColumns ? (
      columns.map((column, index) => {
        // If the column is an object with cell and props, wrap in Th
        if (isThObject(column)) {
          return (
            <Th key={index} {...column.props} data-ouia-component-id={`${ouiaId}-th-${index}`}>
              {column.cell}
            </Th>
          );
        }
        // If the column is already a React element (like <Th>), render it directly
        if (isReactElement(column)) {
          return column;
        }
        // Otherwise, wrap the content in Th
        return (
          <Th key={index} data-ouia-component-id={`${ouiaId}-th-${index}`}>
            {column}
          </Th>
        );
      })
    ) : (
      [ ...Array(rowCellsCount) ].map((_, index) => (
        <Th key={index} data-ouia-component-id={`${ouiaId}-th-${index}`}>
          <Skeleton />
        </Th>
      ))
    ))
  ]
  , [ hasCustomColumns, isExpandable, isSelectable, isTreeTable, columns, rowCellsCount, ouiaId ]);
  

  return (
    <Thead data-ouia-component-id={`${ouiaId}-thead`} {...rest}>
      <Tr ouiaId={`${ouiaId}-tr-head`}>
        {cells}
      </Tr>
    </Thead>
  );
};

export default SkeletonTableHead;
