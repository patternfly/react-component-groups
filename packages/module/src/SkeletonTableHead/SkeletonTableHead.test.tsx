import { render } from '@testing-library/react';
import SkeletonTableHead from './SkeletonTableHead';
import { Table, Th } from '@patternfly/react-table';

describe('SkeletonTableHead component', () => {
  it('should render correctly with count', () => {
    expect(render(<Table><SkeletonTableHead columnsCount={2} isSelectable isExpandable /></Table>)).toMatchSnapshot();
  });

  it('should render correctly with string columns', () => {
    expect(render(<Table><SkeletonTableHead columns={[ 'First', 'Second' ]} isTreeTable isSelectable /></Table>)).toMatchSnapshot();
  });

  it('should render correctly with Th element columns without nesting', () => {
    const { container } = render(
      <Table>
        <SkeletonTableHead
          columns={[
            <Th key="1" sort={{ columnIndex: 0, sortBy: {} }}>First</Th>,
            <Th key="2" sort={{ columnIndex: 1, sortBy: {} }}>Second</Th>
          ]}
        />
      </Table>
    );

    // Ensure there are no nested <th> elements
    const theadCells = container.querySelectorAll('thead th');
    theadCells.forEach(th => {
      const nestedTh = th.querySelector('th');
      expect(nestedTh).toBeNull();
    });

    expect(container).toMatchSnapshot();
  });
});
