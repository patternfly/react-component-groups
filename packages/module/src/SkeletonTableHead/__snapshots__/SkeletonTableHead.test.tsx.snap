// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SkeletonTableHead component should render correctly with Th element columns without nesting 1`] = `
<div>
  <table
    class="pf-v6-c-table pf-m-grid-md"
    data-ouia-component-id="OUIA-Generated-Table-3"
    data-ouia-component-type="PF6/Table"
    data-ouia-safe="true"
    role="grid"
  >
    <thead
      class="pf-v6-c-table__thead"
      data-ouia-component-id="SkeletonTableHeader-thead"
    >
      <tr
        class="pf-v6-c-table__tr"
        data-ouia-component-id="SkeletonTableHeader-tr-head"
        data-ouia-component-type="PF6/TableRow"
        data-ouia-safe="true"
      >
        <th
          class="pf-v6-c-table__th pf-v6-c-table__sort"
          scope="col"
          tabindex="-1"
        >
          <button
            class="pf-v6-c-table__button"
            type="button"
          >
            <div
              class="pf-v6-c-table__button-content"
            >
              <span
                class="pf-v6-c-table__text"
              >
                First
              </span>
              <span
                class="pf-v6-c-table__sort-indicator"
              >
                <svg
                  aria-hidden="true"
                  class="pf-v6-svg"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  viewBox="0 0 256 512"
                  width="1em"
                >
                  <path
                    d="M214.059 377.941H168V134.059h46.059c21.382 0 32.09-25.851 16.971-40.971L144.971 7.029c-9.373-9.373-24.568-9.373-33.941 0L24.971 93.088c-15.119 15.119-4.411 40.971 16.971 40.971H88v243.882H41.941c-21.382 0-32.09 25.851-16.971 40.971l86.059 86.059c9.373 9.373 24.568 9.373 33.941 0l86.059-86.059c15.12-15.119 4.412-40.971-16.97-40.971z"
                  />
                </svg>
              </span>
            </div>
          </button>
        </th>
        <th
          class="pf-v6-c-table__th pf-v6-c-table__sort"
          scope="col"
          tabindex="-1"
        >
          <button
            class="pf-v6-c-table__button"
            type="button"
          >
            <div
              class="pf-v6-c-table__button-content"
            >
              <span
                class="pf-v6-c-table__text"
              >
                Second
              </span>
              <span
                class="pf-v6-c-table__sort-indicator"
              >
                <svg
                  aria-hidden="true"
                  class="pf-v6-svg"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  viewBox="0 0 256 512"
                  width="1em"
                >
                  <path
                    d="M214.059 377.941H168V134.059h46.059c21.382 0 32.09-25.851 16.971-40.971L144.971 7.029c-9.373-9.373-24.568-9.373-33.941 0L24.971 93.088c-15.119 15.119-4.411 40.971 16.971 40.971H88v243.882H41.941c-21.382 0-32.09 25.851-16.971 40.971l86.059 86.059c9.373 9.373 24.568 9.373 33.941 0l86.059-86.059c15.12-15.119 4.412-40.971-16.97-40.971z"
                  />
                </svg>
              </span>
            </div>
          </button>
        </th>
      </tr>
    </thead>
  </table>
</div>
`;

exports[`SkeletonTableHead component should render correctly with count 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <table
        class="pf-v6-c-table pf-m-grid-md"
        data-ouia-component-id="OUIA-Generated-Table-1"
        data-ouia-component-type="PF6/Table"
        data-ouia-safe="true"
        role="grid"
      >
        <thead
          class="pf-v6-c-table__thead"
          data-ouia-component-id="SkeletonTableHeader-thead"
        >
          <tr
            class="pf-v6-c-table__tr"
            data-ouia-component-id="SkeletonTableHeader-tr-head"
            data-ouia-component-type="PF6/TableRow"
            data-ouia-safe="true"
          >
            <th
              class="pf-v6-c-table__th"
              scope="col"
              tabindex="-1"
            >
              <span
                class="pf-v6-u-screen-reader"
              >
                Data expansion table header cell
              </span>
            </th>
            <th
              class="pf-v6-c-table__th"
              scope="col"
              tabindex="-1"
            >
              <span
                class="pf-v6-u-screen-reader"
              >
                Data selection table header cell
              </span>
            </th>
            <th
              class="pf-v6-c-table__th"
              data-ouia-component-id="SkeletonTableHeader-th-0"
              scope="col"
              tabindex="-1"
            >
              <div
                class="pf-v6-c-skeleton"
              >
                <span
                  class="pf-v6-screen-reader"
                />
              </div>
            </th>
            <th
              class="pf-v6-c-table__th"
              data-ouia-component-id="SkeletonTableHeader-th-1"
              scope="col"
              tabindex="-1"
            >
              <div
                class="pf-v6-c-skeleton"
              >
                <span
                  class="pf-v6-screen-reader"
                />
              </div>
            </th>
          </tr>
        </thead>
      </table>
    </div>
  </body>,
  "container": <div>
    <table
      class="pf-v6-c-table pf-m-grid-md"
      data-ouia-component-id="OUIA-Generated-Table-1"
      data-ouia-component-type="PF6/Table"
      data-ouia-safe="true"
      role="grid"
    >
      <thead
        class="pf-v6-c-table__thead"
        data-ouia-component-id="SkeletonTableHeader-thead"
      >
        <tr
          class="pf-v6-c-table__tr"
          data-ouia-component-id="SkeletonTableHeader-tr-head"
          data-ouia-component-type="PF6/TableRow"
          data-ouia-safe="true"
        >
          <th
            class="pf-v6-c-table__th"
            scope="col"
            tabindex="-1"
          >
            <span
              class="pf-v6-u-screen-reader"
            >
              Data expansion table header cell
            </span>
          </th>
          <th
            class="pf-v6-c-table__th"
            scope="col"
            tabindex="-1"
          >
            <span
              class="pf-v6-u-screen-reader"
            >
              Data selection table header cell
            </span>
          </th>
          <th
            class="pf-v6-c-table__th"
            data-ouia-component-id="SkeletonTableHeader-th-0"
            scope="col"
            tabindex="-1"
          >
            <div
              class="pf-v6-c-skeleton"
            >
              <span
                class="pf-v6-screen-reader"
              />
            </div>
          </th>
          <th
            class="pf-v6-c-table__th"
            data-ouia-component-id="SkeletonTableHeader-th-1"
            scope="col"
            tabindex="-1"
          >
            <div
              class="pf-v6-c-skeleton"
            >
              <span
                class="pf-v6-screen-reader"
              />
            </div>
          </th>
        </tr>
      </thead>
    </table>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`SkeletonTableHead component should render correctly with string columns 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <table
        class="pf-v6-c-table pf-m-grid-md"
        data-ouia-component-id="OUIA-Generated-Table-2"
        data-ouia-component-type="PF6/Table"
        data-ouia-safe="true"
        role="grid"
      >
        <thead
          class="pf-v6-c-table__thead"
          data-ouia-component-id="SkeletonTableHeader-thead"
        >
          <tr
            class="pf-v6-c-table__tr"
            data-ouia-component-id="SkeletonTableHeader-tr-head"
            data-ouia-component-type="PF6/TableRow"
            data-ouia-safe="true"
          >
            <th
              class="pf-v6-c-table__th"
              data-ouia-component-id="SkeletonTableHeader-th-0"
              scope="col"
              tabindex="-1"
            >
              First
            </th>
            <th
              class="pf-v6-c-table__th"
              data-ouia-component-id="SkeletonTableHeader-th-1"
              scope="col"
              tabindex="-1"
            >
              Second
            </th>
          </tr>
        </thead>
      </table>
    </div>
  </body>,
  "container": <div>
    <table
      class="pf-v6-c-table pf-m-grid-md"
      data-ouia-component-id="OUIA-Generated-Table-2"
      data-ouia-component-type="PF6/Table"
      data-ouia-safe="true"
      role="grid"
    >
      <thead
        class="pf-v6-c-table__thead"
        data-ouia-component-id="SkeletonTableHeader-thead"
      >
        <tr
          class="pf-v6-c-table__tr"
          data-ouia-component-id="SkeletonTableHeader-tr-head"
          data-ouia-component-type="PF6/TableRow"
          data-ouia-safe="true"
        >
          <th
            class="pf-v6-c-table__th"
            data-ouia-component-id="SkeletonTableHeader-th-0"
            scope="col"
            tabindex="-1"
          >
            First
          </th>
          <th
            class="pf-v6-c-table__th"
            data-ouia-component-id="SkeletonTableHeader-th-1"
            scope="col"
            tabindex="-1"
          >
            Second
          </th>
        </tr>
      </thead>
    </table>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
